src/Features/CSharp/Portable/ConvertLinq/CSharpConvertLinqQueryToForEachProvider.cs

```diff
@@ -0,0 +1,608 @@
+                documentUpdate = default;
+                return false;
+            }
+
+            private bool TryGetTypeSyntax(ExpressionSyntax expression, out TypeSyntax typeSyntax)
+            {
+                var typeSymbol = _semanticModel.GetTypeInfo(expression, _cancellationToken).Type;
+                if (typeSymbol.TypeKind == TypeKind.Error ||
+                    typeSymbol.ContainsAnonymousType())
+                {
+                    typeSyntax = default;
+                    return false;
+                }
+
+                typeSyntax = typeSymbol.GenerateTypeSyntax();
+                return true;
+            }
+
+            private string GetFreeSymbolName(string prefix, int position)
+                => Shared.Utilities.NameGenerator.GenerateUniqueName(
+                    prefix,
+                    _semanticModel.LookupSymbols(position).Select(symbol => symbol.Name).ToImmutableHashSet(),
+                    StringComparer.CurrentCulture);
+
+            private bool TryConvertIfInForEach(QueryExpressionSyntax source, ForEachStatementSyntax forEachStatement, out DocumentUpdate documentUpdate)
+            {
+                var statement = forEachStatement.Statement;
+                StatementSyntax internalNodeMethod(ExpressionSyntax expression)
+                {
+                    if (expression is IdentifierNameSyntax identifier &&
+                        identifier.Identifier.ValueText == forEachStatement.Identifier.ValueText)
+                    {
+                        return statement;
+                    }
+                    else
+                    {
+                        var declaration = SyntaxFactory.LocalDeclarationStatement(
+                            SyntaxFactory.VariableDeclaration(
+                                VarNameIdentifier,
+                                SyntaxFactory.SingletonSeparatedList(
+                                    SyntaxFactory.VariableDeclarator(
+                                        forEachStatement.Identifier,
+                                        argumentList: null,
+                                        SyntaxFactory.EqualsValueClause(expression)))));
+                        return AddToBlockTop(declaration, statement);
+                    }
+                }
+
+                if (TryConvertQueryExpression(source, internalNodeMethod, out StatementSyntax newNode))
+                {
+                    documentUpdate = new DocumentUpdate(forEachStatement, newNode);
+                    return true;
+                }
+
+                documentUpdate = default;
+                return false;
+            }
+
+            private bool TryConvertIfInReturnStatement(QueryExpressionSyntax source, ReturnStatementSyntax returnStatement, out DocumentUpdate documentUpdate)
+            {
+                StatementSyntax internalNodeMethod(ExpressionSyntax expression) => SyntaxFactory.YieldStatement(SyntaxKind.YieldReturnStatement, expression);
```

Comment:
nit: consider putting => on next line (indented).  it will make it clearer these are local functions.